src/varnish.c

```diff
@@ -127,36 +134,106 @@ static int varnish_submit_gauge(const char *plugin_instance, /* {{{ */
+
+  return varnish_submit(plugin_instance, category, target, type, type_instance,
+                        (value_t){
+                            .gauge = (gauge_t)gauge_value,
+                        });
+} /* }}} int varnish_submit_gauge_with_target */
+
 static int varnish_submit_derive(const char *plugin_instance, /* {{{ */
                                  const char *category, const char *type,
                                  const char *type_instance,
                                  uint64_t derive_value) {
-  return varnish_submit(plugin_instance, category, type, type_instance,
+  return varnish_submit(plugin_instance, category, NULL, type, type_instance,
                         (value_t){
                             .derive = (derive_t)derive_value,
                         });
 } /* }}} int varnish_submit_derive */
 
+static int
+varnish_submit_derive_with_target(const char *plugin_instance, /* {{{ */
+                                  const char *category, const char *target,
+                                  const char *type, const char *type_instance,
+                                  uint64_t derive_value) {
+  return varnish_submit(plugin_instance, category, target, type, type_instance,
+                        (value_t){
+                            .derive = (derive_t)derive_value,
+                        });
+} /* }}} int varnish_submit_derive_with_target */
+
 #if HAVE_VARNISH_V3 || HAVE_VARNISH_V4 || HAVE_VARNISH_V5 || HAVE_VARNISH_V6
 static int varnish_monitor(void *priv,
                            const struct VSC_point *const pt) /* {{{ */
 {
   uint64_t val;
   const user_config_t *conf;
   const char *name;
+  //  const char *stat_category;
+  const char *stat_target;
 
   if (pt == NULL)
     return 0;
 
   conf = priv;
 
-#if HAVE_VARNISH_V5 || HAVE_VARNISH_V6
+#if HAVE_VARNISH_V6
+  /*
+   stats examples:
+   MAIN.threads
+   SMA.s0.c_req
+   SMA.Transient.c_req
+   VBE.vclName.backendName.req // where vclName changes every time a new vcl is
+   loaded
+  */
+  const int MAX_TOKENS = 4; // we expect 4 maximum tokens
+  char *token_array[MAX_TOKENS];
+  int i = 0;
+  int token_count = -1;
+  char *buffer = strdup(pt->name);
+  char *running = buffer;
+  do {
+    token_array[i] = strsep(&running, ".");
+    if (token_array[i] == NULL) {
+      token_count = i - 1;
+    }
+    i++;
+  } while (token_count == -1 &&
+           i <= MAX_TOKENS); // doing one more round (thanks to lte) to ensure
+                             // next token is NULL
+  if (token_count == -1) {
+    free(buffer);
+    return EINVAL;
+  }
+
+  char namebuff[DATA_MAX_NAME_LEN];
+  char targetbuff[DATA_MAX_NAME_LEN];
+
+  sstrncpy(namebuff, token_array[token_count], sizeof(namebuff));
+  name = namebuff;
+
+  // stat_category is available at token_array[0], if we ever get stats with the
+  // same name for different categories;
+
+  if (token_count > 1) {
+    sstrncpy(targetbuff, token_array[token_count - 1], sizeof(targetbuff));
```

Comment:
This line and the comment above contradict each other. Which one is it, first element of `token_array` (`token_array[0]`) or third-to-last element, i.e. `token_array[-3]`?

Reply:
comment above is to indicate the first token is a category, and can be used later. 
here we're computing second to last element to get the stat target (ie s0, transient, or a disk name).  I should add a comment to better separate, or remove the previous comment completely